### PR TITLE
Store modification & deletion dates for series and add range-lookup method to `SeriesService`

### PR DESCRIPTION
**Pull request by LukasKalbertodt**
Like other PRs before, these changes are required by the upcoming Tobira module. The module will enable the new video portal Tobira to synchronize with Opencast somewhat efficiently. It's still in development in https://github.com/elan-ev/opencast-tobira/. We want to push all tobira-unrelated improvements to the community as early as possible though. Hence this PR.

---

To synchronized with Opencast efficiently, Opencast needs to know when a "thing" last changed. That was already true for events (at least in the search service), but Opencast did not know this information for series. This PR changes that by adding a modified date field to series. 

To notify other applications, like Tobira, about a deleted series, we now store a deletion date instead of actually deleting the series.

To use this new information, a new method is added to the `SeriesService` that allows modification date range lookup and returns series sorted by modification date. Unfortunately, to implement the remote impl, we needed to add a new JSON API as well. This is described as internal and is not expected to be used outside of Opencast itself. 

Also see commit message for more details. This PR can probably be reviewed commit-by-commit fairly well.

### Your pull request should…

* [ ] have a concise title
* [ ] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [ ] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated tests
* [ ] have a clean commit history
* [ ] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

### CODE REVIEW

File: modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java
Comment (mtneug) [inline modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java:419]: Does this mean the Tobira API user needs to be (org) admin? I would not give Tobira global admin rights. Org admin maybe, but I'm still unsure if this is a good design. 

Do you expect this method to be used internally or can we simply rely on Spring security for the REST endpoint?

Comment (LukasKalbertodt) [reply modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java:419]: Currently, we need the Tobira user to be org admin, yes. I was also not happy with this, but ultimately, Tobira needs read access to all events that should potentially appear on the Tobira page. Further, it needs write access to all events that could potentially modified through Tobira. For most institutions that's often simply "Tobira needs read and write access to all videos". After all, users of Tobira perform actions in Opencast *through* Tobira. So it kind of makes sense. This is the same for some other integrations, e.g. the Ilias plugin which also requires a very powerful Opencast user.

Of course, you could always add a special Tobira role to all ACLs to avoid handing out org admin rights. 

Your last quesiton I don't quite understand. I expect the REST endpoint to be used *only* by the series service remote impl.

Comment (mtneug) [reply modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java:419]: This is not true. ILIAS and Moodle don't require org admin rights and you can thus have videos that are not seen by those systems.

Comment (LukasKalbertodt) [reply modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java:419]: I just said "a very powerful Opencast user", not org admin. Though yes, that's still not technically correct because the user used by Ilias/Moodle does not *need* to have lots of privileges. But it typically does, i.e. is allowed to read/write the majority (if not all) events.

Again, I also don't like the Tobira module requiring org admin right now. And I would like to change that. I think this can be easily changed in the future. I'm just saying the OC user used by Tobira has very similar requirements to OC users used by Ilias/Moodle.

**Comment by LukasKalbertodt** [comment]
This is still missing an appropriate SQL migration. And I noticed that the code likely has to be adjusted to deal with old series that don't have a valid modification date. So I will work on this and mark as "read for review" as soon as it's done.

File: modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java
Comment (mtneug) [reply modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java:419]: Fair enough and I think we see this similarly :)

File: modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesEntity.java
Comment (Arnei) [inline modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesEntity.java:67]: Could you explain to me when do we have a series that is deleted still in the db? As far as I know, deleted series are removed from the db, so storing a deletionDate makes little sense?

Comment (LukasKalbertodt) [reply modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesEntity.java:67]: This PR also changes that. When deleting a series, it is not actually deleted from the DB, but we only set the deleted date to a non-null value.

```diff
- em.remove(entity);
+ entity.setModifiedDate(new Date());
+ entity.setDeletionDate(new Date());
+ em.merge(entity);
```

https://github.com/opencast/opencast/pull/3076/commits/afb82d27d61370bb2d5903e4a25881c9aad11401#diff-8ebbecff33097385ffb1bdcc22e6a8acd012f150472562fa17ab98838c7afd43L172-L173

Comment (Arnei) [reply modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesEntity.java:67]: Right, not sure how I missed that. Thanks for the link

Comment (Arnei) [reply modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesEntity.java:67]: Does that mean there is no way of permanently deleting a series from the db anymore short of directly removing them from the db by hand?

Comment (LukasKalbertodt) [reply modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesEntity.java:67]: Yes, that is correct. I mean, for Opencast, it is deleted as all SQL queries retrieving series now explicitly check for `deletion_date is null`. But yes, the data is still in the database. 

Maybe instead of just setting the modification and deletion dates on removal, all other columns should be set to empty or dummy values? But that could always be changed later I guess?

**Review by Arnei — APPROVED** [review]

**Comment by Arnei** [comment]
Looks good to me, but I think I'd like a second opinion here. Particularly on the "series aren't actually deleted anymore"-part.  So far, I've not actually managed to come up with a good argument on why this would be a problem, but I still feel like it could be a problem. Maybe there are some database limitations or weird data privacy laws I'm unaware of.

File: docs/upgrade/10_to_11/postgresql9.sql
Comment (lkiesow) [inline docs/upgrade/10_to_11/postgresql9.sql:23]: Does it make sense to run at the end:

```sql
VACUUM (VERBOSE, ANALYZE) oc_series;
```

File: modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/DublinCoreGsonAdapter.java
Comment (lkiesow) [inline modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/DublinCoreGsonAdapter.java:29]: If you use Gson, you probably also want to use Gson's exception type:
[`com.google.gson.JsonParseException](https://www.javadoc.io/doc/com.google.code.gson/gson/2.6.2/com/google/gson/JsonParseException.html)

File: modules/series-service-impl/src/main/java/org/opencastproject/series/endpoint/SeriesRestService.java
Comment (lkiesow) [inline modules/series-service-impl/src/main/java/org/opencastproject/series/endpoint/SeriesRestService.java]: Gson is thread-safe, and you can make this a class field. Constructing this every time has quite some overhead and caused problems for Opencast at several places.

File: modules/series-service-impl/src/main/java/org/opencastproject/series/endpoint/SeriesRestService.java
Comment (lkiesow) [inline modules/series-service-impl/src/main/java/org/opencastproject/series/endpoint/SeriesRestService.java]: This lets users spam the logs with warnings.
For a client error, the error is returned.
Make this a debug log:
```suggestion
    logger.debug("Bad request to /series/allInRangeAdministrative: {}", msg);
```

File: modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java
Comment (lkiesow) [inline modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java]: Maybe make sure they are actually identical?
```suggestion
      var date = new Date();
      entity.setModifiedDate(date);
      entity.setDeletionDate(date);
```

File: modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java
Comment (lkiesow) [inline modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java:436]: Above, you check if the user is an admin or an organizational admin. This query returns all events from all organizations. This means, you allow admins from one organization to access events from another organization. I don't think that's intended, is it?

File: modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java
Comment (lkiesow) [inline modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java:443]: `SeriesEntity` holds basically the same data the new `Series` does. Would it make sense to have `SeriesEntity` implement/extend `Series`?

File: modules/series-service-remote/src/main/java/org/opencastproject/series/remote/SeriesServiceRemoteImpl.java
Comment (lkiesow) [inline modules/series-service-remote/src/main/java/org/opencastproject/series/remote/SeriesServiceRemoteImpl.java]: Make these class fields.

**Review by lkiesow — CHANGES_REQUESTED** [review]

File: modules/series-service-impl/src/main/java/org/opencastproject/series/endpoint/SeriesRestService.java
Comment (lkiesow) [inline modules/series-service-impl/src/main/java/org/opencastproject/series/endpoint/SeriesRestService.java:375]:
```suggestion
              defaultValue = "",
```

File: modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java
Comment (LukasKalbertodt) [reply modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java:443]: I don't really think so. `SeriesEntity` for me is just the connection to the DB and I would not want to mix it with `Series`. For example, `Series` contains `DublinCoreCatalog dublinCore`. If `SeriesEntity` would derive from `Series` that would have to be `null` all the time? And from experience, in these situations, sooner or later there will be more and more differences between the "database adapter" and the thing you want to use in your actual code.

File: docs/upgrade/10_to_11/postgresql9.sql
Comment (LukasKalbertodt) [reply docs/upgrade/10_to_11/postgresql9.sql:23]: Not sure to be honest. At least in postgres (as far as I know), the first 8 nullable columns don't have any overhead as there is always an 8bit mask per row for exactly that. But :shrug:

File: modules/series-service-impl/src/main/java/org/opencastproject/series/endpoint/SeriesRestService.java
Comment (LukasKalbertodt) [reply modules/series-service-impl/src/main/java/org/opencastproject/series/endpoint/SeriesRestService.java:375]: Nop, that still shows `0` in the test form box :/

**Comment by LukasKalbertodt** [comment]
@lkiesow I think I fixed all of your remarks (except the four which I comment-replied to). Most changes have been amended to the respective commits. Only the security fix is its own commit.

File: modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/DublinCoreGsonAdapter.java
Comment (LukasKalbertodt) [reply modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/DublinCoreGsonAdapter.java:29]: Well, I am not throwing that exception. I am catching it. Because I'm using an existing method which uses JSON simple. It's a bit stupid but was the easiest solution. Everything else would have required writing lots more code and likely duplicating logic. Kinda annoying to have multiple JSON libraries.

File: modules/series-service-impl/src/main/java/org/opencastproject/series/endpoint/SeriesRestService.java
Comment (lkiesow) [reply modules/series-service-impl/src/main/java/org/opencastproject/series/endpoint/SeriesRestService.java:375]: Interesting. Well… let's ignore that for now.

File: modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java
Comment (lkiesow) [reply modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java:436]: This is now resolved in the query `getAllModifiedSince`

File: modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java
Comment (lkiesow) [reply modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java:443]: I would have probably solved that by making `Series` an interface and just requiring the methods, but fair enough, it's not really important.

**Review by lkiesow — APPROVED** [review]

File: modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java
Comment (KatrinIhler) [inline modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java:324]: Just ran into this while testing my migration scripts for Bern - this doesn't actually seems to work, since deleteSeries() would only mark the entity as deleted again, not remove it (found no method to do this), but mostly because this throws a NotFound for entities which are already marked as deleted. To test this, just create a series, delete it, and then try to recreate a series with the same id. Since you know how this feature is supposed to work, can you take a look at this?

**Pull request merged**
